plugins/SlicerT/SlicerT.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * SlicerT.h - declaration of class SlicerT
+ *
+ * Copyright (c) 2023 Daniel Kauss Serna <daniel.kauss.serna@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef LMMS_SLICERT_H
+#define LMMS_SLICERT_H
+
+#include <algorithm>
+#include <fftw3.h>
+
+#include "AutomatableModel.h"
+#include "Instrument.h"
+#include "InstrumentView.h"
+#include "Note.h"
+#include "SampleBuffer.h"
+#include "SlicerTView.h"
+#include "lmms_basics.h"
+
+namespace lmms {
+
+class PlayBackState
```

Comment:
It is created for every note and stores that notes data. Similar to `SampleBuffer::handlestate`. I do agree with the constructor, so that's done.

Reply:
> It is created for every note and stores that notes data. Similar to SampleBuffer::handlestate. I do agree with the constructor, so that's done.

Ah, okay, then that's fine. However, and this bit of a nitpick, but there are some issues I found with the declaration of this class (naming issues mostly). I would brush up this class and define it like this:

```cpp
class PlaybackState
{
public:
	explicit PlaybackState(float startFrame)
		: m_startFrame(startFrame)
		, m_resamplingState(src_new(SRC_SINC_MEDIUM_QUALITY, DEFAULT_CHANNELS, nullptr))
	{
		if (!m_resamplingState) { throw std::runtime_error{"Failed to create sample rate converter object"}; }
	}

	~PlaybackState() noexcept { src_delete(m_resamplingState); }

	float startFrame() const { return m_startFrame; }
	void setStartFrame(float startFrame) { m_startFrame = startFrame; }

	SRC_STATE* resamplingState() const { return m_resamplingState; }

private:
	float m_startFrame;
	SRC_STATE* m_resamplingState;
};
```

Another problem I should mention (although you don't have to take any action on) is that we should have a single source of truth for the desired resample quality. This codebase seems to just choose it's own resample state everywhere, which creates a lot of inconsistency. The audio engine's resample quality right now I believe is the worst one (zero order hold), so that should eventually be bumped up and resample converter objects should come directly from the engine. Mentioned it here just to remind myself of it and to let you know.

Reply:
Update the class to this, but I changed `startFrame` to `currentNoteDone`. That variable holds how much of the note is done, not where it starts. As for what resampling quality is used, I actually used `SRC_LINEAR` because the sync options created weird artifacts. If you change to `SRC_SINC_MEDIUM_QUALITY` for example and load a long pure sine wave, there are jumps at regular intervals.